### PR TITLE
doc: Extend quarantine doc to cover flaky required lanes

### PR DESCRIPTION
**Pull request by brianmcarey**
### What this PR does

This document update covers what actions should be taken when test lanes are flaking with multiple test failures where quaranting the individual tests might not make sense.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Special notes for your reviewer
/cc @dhiller @EdDev 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

**Comment by brianmcarey** [comment]
/hold 

For mailing list feedback

**Review by acardace — APPROVED** [review]

File: docs/quarantine.md
Comment (EdDev) [inline docs/quarantine.md:123]: Can we add an enforcement action here? Can be a subsection.
How exactly this is enforced? Even if it is not automated/configured, the responsibility should be explicitly defined on the one who us responsible to assure no PRs are merged unless it is fixing the flake (e.g. it could be under the specific SIG responsibility).

**Review by EdDev — APPROVED** [review]
Makes sense to me, thanks!
Hope we will never need something like this.

File: docs/quarantine.md
Comment (brianmcarey) [reply docs/quarantine.md:123]: I was thinking about how this could be enforced - SIG CI could hold PRs if they are not happy that the flakes are being given the appropriate focus by the responsible SIG. 

Something like:
```
If the flaky test lane is not receiving the required attention from the responsible SIG, 
SIG CI will hold merge queue PRs from the responsible SIG that are not related to a fix.
```

File: docs/quarantine.md
Comment (dankenigsberg) [inline docs/quarantine.md:113]: Do we have a clear metric to measure the failure rate?

IMHO 40% is way too high. What is the current failure rate of our worst lane? I don't think we should let a lane that is failing 10% of the time to block the rest of the project.

Comment (brianmcarey) [reply docs/quarantine.md:113]: > Do we have a clear metric to measure the failure rate?

We can search for the flake and see the percentage impact of that flake - for example - https://search.ci.kubevirt.io/?search=log+line+even+trying+to+flood+the+serial+console+for+QOSGuaranteed+VMs&maxAge=336h&context=1&type=bug%2Bissue%2Bjunit&name=&excludeName=&maxMatches=5&maxBytes=20971520&groupBy=job

> 
> IMHO 40% is way too high. What is the current failure rate of our worst lane? I don't think we should let a lane that is failing 10% of the time to block the rest of the project.

Yes I think it is a bit high too - I was working off old numbers in my head where we used to have an average failure rate of about 20% - we are now averaging between ~ 9% - 13% failure rate overall e2e lanes[1]. 25% failure rate is a better limit as that is when it will begin to impact our ability to merge PRs. 

[1]https://grafana.ci.kubevirt.io/d/efpTS3t4z/e2e-jobs-overview-v2?orgId=1&viewPanel=13&from=1722294422579&to=1723452234415

**Comment by brianmcarey** [comment]
@EdDev @acardace could you please take a look again? thanks

File: docs/quarantine.md
Comment (dankenigsberg) [reply docs/quarantine.md]: Thanks. Would you include a short explanation how I can tell what is the failure rate of a specific lane, or how I can find those lanes that fail more often than 25%?
It would help me, and I think also others, understand when it's time to downgrade a lane.

File: docs/quarantine.md
Comment (xpivarc) [inline docs/quarantine.md:116]: Delivering a fix for a one flake is sometimes a job for days not hours. I am curious how do we set some deadline.

Let's say we encounter clustered failures. What do we gain by  merging PRs/quarantine? 
Imho if we continue merging PRs with a blocking lane we risk that something new will fall through our gating, potentially extending the issue.  On the other hand, not merging PRs for few hours/days, assuming we have people urgently attending, we don't risk new failures being introduced. One obvious negative is to not be able to merge fixes for stable branches but here we can simply override. Other than that we still have an option to backtrack what PRs might have cause this and revert them.

So I am wondering if we are setting up our self for failure here.

On the other hand I believe we need to close the loop on the existing quarantine definition, where we don't make sure these tests are being worked on with priority. We continue releasing without potential valuable feedback from them.

Comment (EdDev) [reply docs/quarantine.md:116]: I think the current proposed text highlights that unstable production or test code should not slow down the whole project. If a specific SIG or subproject suffers from too much instability, the contributors as a whole should re-prioritize their efforts to solve the problem over anything else. One way to do it is by blocking any new changes to get in, solving exactly the cases mentioned in the arguments (e.g. letting bugs get in because it is not covered).

> Other than that we still have an option to backtrack what PRs might have cause this and revert them.

I think this is one of the best solutions for an immediate action.
For new tests or new production code that can be recognized as causing the problem, it is this or a real fix, whichever comes first.

> So I am wondering if we are setting up our self for failure here.

One can argue that it will just expose an already existing failure, making it visible and not hidden under the hood.

> On the other hand I believe we need to close the loop on the existing quarantine definition, where we don't make sure these tests are being worked on with priority. We continue releasing without potential valuable feedback from them.

I agree. I would expect the quarantined tests owners to be be responsible for this. If they do not have the time to maintain it, then the tests should be removed after a period of time. There is no choice but to balance between our hope to cover as much as possible and our ability to maintain all the instability of such tests (i.e. to fix them).

Comment (brianmcarey) [reply docs/quarantine.md:116]: > Delivering a fix for a one flake is sometimes a job for days not hours. I am curious how do we set some deadline.

Exactly if its going to take longer than a day to fix then we know that we shouldn't leave the lane in question required as it is going to block other SIGs.

> 
> Let's say we encounter clustered failures. What do we gain by merging PRs/quarantine? Imho if we continue merging PRs with a blocking lane we risk that something new will fall through our gating, potentially extending the issue. 

I would see the risk of breakage as an added incentive for the responsible SIG to make this a top priority as it should be.  The responsible SIG should really not be merging PRs related to the flaking test lane during this quarantine period which should lower the risk of breaking something. 

Again these scenarios are relatively rare but the aim of this procedure is to ensure that its all hand on deck from the responsible SIG to fix the broken lane.

>On the other hand, not merging PRs for few hours/days, assuming we have people urgently attending, we don't risk new failures being introduced. One obvious negative is to not be able to merge fixes for stable branches but here we can simply override. Other than that we still have an option to backtrack what PRs might have cause this and revert them.
> 
> So I am wondering if we are setting up our self for failure here.
> 
> On the other hand I believe we need to close the loop on the existing quarantine definition, where we don't make sure these tests are being worked on with priority. We continue releasing without potential valuable feedback from them.

Comment (dankenigsberg) [reply docs/quarantine.md:116]: > Delivering a fix for a one flake is sometimes a job for days not hours. I am curious how do we set some deadline.

It may be hard to set a deadline, but we must have one. Let us assume that the windows lane is suddenly failing 90% of the times, making PRs merge after ~10 retries and many hours. @xpivarc for how long would you say we should accept this? Two days? A week? A month?

I recall having the fossa test fail 100% of the time. We attempted to fix it for a few days, failed to do so, so converted it to non-voting. Sure, this means that non-free code can sneak in; but the alternative is blocking everyone else.

We need this policy to declare the sweet point that we follow. Hopefully, as we improve our tests, we can tighten the two parameters (probability of failure and expected resolution time).

File: docs/quarantine.md
Comment (dankenigsberg) [inline docs/quarantine.md:119]: Please excuse me, but I don't know what the relevant error may be. Can we add an example, e.g. how can I find the current failure rate of the windows lane?

Comment (dankenigsberg) [reply docs/quarantine.md:119]: ah, sorry. I did not notice that this paragraph is about finding occurrence of a flaky *test*, not a flaky lane.

Would you add the windows example to the bulllet point in https://github.com/kubevirt/kubevirt/pull/12494/files#diff-4ff5f85030a93d12b9e7929720bd3d43b5f73cba313b3e98a56dbefe2d2abfabR113 ?

I'd like to add an example in order to make it easier for someone like myself to find the current failure rate of a lane that seems problematic.

BTW, does this statistic include cases where the lane failed justly, because it was executed on a bad PR?

Comment (brianmcarey) [reply docs/quarantine.md:119]: > ah, sorry. I did not notice that this paragraph is about finding occurrence of a flaky _test_, not a flaky lane.

This logic can be applied to a flaky lane as well. 
> 
> Would you add the windows example to the bulllet point in https://github.com/kubevirt/kubevirt/pull/12494/files#diff-4ff5f85030a93d12b9e7929720bd3d43b5f73cba313b3e98a56dbefe2d2abfabR113 ?
> 
> I'd like to add an example in order to make it easier for someone like myself to find the current failure rate of a lane that seems problematic.
> 
> BTW, does this statistic include cases where the lane failed justly, because it was executed on a bad PR?

This statistic includes all failures for the lane so it is not fair to use this one for this case - thats where the CI search comes in to provide an accurate measure of the impact percentage.

File: docs/quarantine.md
Comment (dhiller) [inline docs/quarantine.md:113]: I'd suggest to add a link to where to look that percentage up.

File: docs/quarantine.md
Comment (dhiller) [inline docs/quarantine.md:114]: Maybe rather use a percentage here - i.e. for test lanes with very small numbers of tests (vgpu) this amount might never be reached.

**Review by dhiller — COMMENTED** [review]
Thanks @brianmcarey - great addition to the quarantining procedure!

I have made two comments inline.

File: docs/quarantine.md
Comment (dankenigsberg) [reply docs/quarantine.md:113]: +1

We can even make use of the new https://github.com/kubevirt/ci-health?tab=readme-ov-file#failures-per-sig-against-last-code-push-for-merged-prs

File: docs/quarantine.md
Comment (brianmcarey) [reply docs/quarantine.md:114]: 10 failures in the sig-storage lane is 3.24% while 10 failures in the sig-operator lane is 18.2% - due to the wide variety in the number of tests run per lane I'm not sure a percentage would be the best fit here. 

This is just to qualify it as too many tests to quarantine individually.

File: docs/quarantine.md
Comment (brianmcarey) [reply docs/quarantine.md:113]: > I'd suggest to add a link to where to look that percentage up.

I suggested using the CI search to calculate the impact of the flake over the last 24 hours 
https://github.com/kubevirt/kubevirt/pull/12494/files#diff-4ff5f85030a93d12b9e7929720bd3d43b5f73cba313b3e98a56dbefe2d2abfabR118

Comment (brianmcarey) [reply docs/quarantine.md:113]: > +1
> 
> We can even make use of the new https://github.com/kubevirt/ci-health?tab=readme-ov-file#failures-per-sig-against-last-code-push-for-merged-prs

These counts are for the last 7 days and only apply to merged PRs so it doesn't show flakes that open PRs in the queue may be hitting.

Comment (dankenigsberg) [reply docs/quarantine.md:113]: > I am not sure if they are the best fit for this.

Unlike counting failures of open PRs, your kubevirt-ci metric is not polluted by bad code. An open PR may be failing for a very good reason - the new PR is bad.

If a lane is so bad that it does not let you merge PRs at all, it would be easy to notice. I think that our biggest problem these days is when a PR (modifying either test or production code or both) ends up increasing failure rate by 10%. I human developer would likely not notice that before a PR is merged, even by searching for failures.
However, your new metric would show a clear bump after a week.

Yes, it is slow, but it is undeniable, and can help us call out the relevant SIG to fix the issue. I worry that the semi-manual search for errors require too much explanation and negotiation.

Maybe the policy can mention both: *either* you find to many failures in error logs *or* the lane metrics jumps above 25%? Either way, the lane should be marked non-voting until fixed by the relevant SIG.

File: docs/quarantine.md
Comment (xpivarc) [reply docs/quarantine.md:116]: Couple of points. Imho this policy in question is totally context depended. Some lanes are more critical than others, sometime a lane can fail only on specific Kubernetes version (or any dependency per se).  It also depend on why the lane fails. Sometimes it can be out of hands of developers, me remembering vfio bug we discovered that lead to the vgpu lane being unusable.

We are talking about blocking but are we really blocking anyone? Our release process last for ~3 months, not merging any PRs for 3 working days would not hurt any group or individual, imho. Of course in case of bug/security fix we are talking about different context and there maintainer can decide after throughout  considerations to override the context.

Rather than defining a strict policy of quarantining a whole lane I ask to think about process that could help me (as gate keeper but others as well) and contributors to guide through such a rare crisis.

File: docs/quarantine.md
Comment (brianmcarey) [reply docs/quarantine.md:113]: Yes fair points - I've added a link to the top failed jobs list - https://github.com/kubevirt/kubevirt/pull/12494/commits/b083974607c80ac6000d54b947eb33001ec594c0

File: docs/quarantine.md
Comment (brianmcarey) [reply docs/quarantine.md:116]: > Couple of points. Imho this policy in question is totally context depended. Some lanes are more critical than others, sometime a lane can fail only on specific Kubernetes version (or any dependency per se). It also depend on why the lane fails. Sometimes it can be out of hands of developers, me remembering vfio bug we discovered that lead to the vgpu lane being unusable.

Yes I agree with you - the context of the failures is very important. SIG CI would take this context into consideration and discuss with the responsible SIG before taking any action. This doc update is to act as a guideline for SIG CI.


> 
> We are talking about blocking but are we really blocking anyone? Our release process last for ~3 months, not merging any PRs for 3 working days would not hurt any group or individual, imho. Of course in case of bug/security fix we are talking about different context and there maintainer can decide after throughout considerations to override the context.

3 days without merging could lead to say ~30 PRs in the queue - this can lead to its own problems with CI load and unnecessary retests but also a large number of different combinations of batches runs being created. Its just wasteful. 

Also it can be frustrating for the other SIGs trying to deliver features unrelated to the flake. 

> 
> Rather than defining a strict policy of quarantining a whole lane I ask to think about process that could help me (as gate keeper but others as well) and contributors to guide through such a rare crisis.

Comment (dankenigsberg) [reply docs/quarantine.md:116]: @xpivarc I agree that the context is important; I tend to agree with you that 3 days may be two short, particularly if it is a weekend. However where would *you* draw the line?

Let us assume that the windows lane is suddenly failing 90% of the times, making PRs merge after ~10 retries and many hours. For how long would you say we should accept this, if it's more than three days - a week? A month?

Every day a PR is not merged due to a reason unrelated to itself (say, a faulty windows lane), we lose the attention of the PR contributor, cause frustration and of course economic cost. At some point, SIG CI should put its foot down and force the owners of the faulty lane to have it fixed ASAP.

File: docs/quarantine.md
Comment (dankenigsberg) [reply docs/quarantine.md:119]: To me this is a strong reason to use the your new metric. For example, we currently have kubevirt-e2e-kind-sriov failing 51 times out of 100 attempts over the recent week https://camo.githubusercontent.com/ddcf87891214d0984a90a503ac3e4f47f264c9fdc80d78f793332c40d72e9443/68747470733a2f2f6b756265766972742e696f2f63692d6865616c74682f6f75747075742f6b756265766972742f6b756265766972742f6661696c65646a6f62312e737667
I think that @xpivarc would agree that this is not acceptable. It is unfair for everyone and sig-network should get to the bottom of it ASAP.  I find it better to base our policy on clear and objective metric.

File: docs/quarantine.md
Comment (EdDev) [reply docs/quarantine.md:127]: IMO the decision should be of SIG-CI or the test owners, but the implementation should be on the SIG that owns the tests. If they fail, only then the SIG-CI should take an extreme action and start actively blocking those PRs. If the SIG-CI take a clocking action, it is an extreme and I would consider the SIG owning the tests as not confirming to the policy.

File: docs/quarantine.md
Comment (xpivarc) [reply docs/quarantine.md:119]: I agree 51 out of 100 is crazy.

File: docs/quarantine.md
Comment (xpivarc) [reply docs/quarantine.md]: Just to understand, now individual tests are failing and they make up to 20% of failure rate for a lane, right? So what do we want to target here?

File: docs/quarantine.md
Comment (xpivarc) [reply docs/quarantine.md:116]: So I would suggest to have a week as period of time where we acknowledge the problem, relevant sig is informed and is making it priority to fix the issues. After the week I would opt-in to quarantine. I am wondering what and if anything can be done to ensure quality while the lane is quarantined.

What do we think?

File: docs/quarantine.md
Comment (xpivarc) [reply docs/quarantine.md:119]: +1 to define on what bases the decision to quarantine lane will take place.

File: docs/quarantine.md
Comment (brianmcarey) [reply docs/quarantine.md:116]: > So I would suggest to have a week as period of time where we acknowledge the problem, relevant sig is informed and is making it priority to fix the issues. After the week I would opt-in to quarantine. I am wondering what and if anything can be done to ensure quality while the lane is quarantined.
> 
> What do we think?

OK a week period seems fair. This will allow us to use the test lane badges in ci-health as a metric on whether to quarantine or not. I will make updates to capture this.

**Review by dhiller — APPROVED** [review]

**Review by xpivarc — APPROVED** [review]

**Pull request merged**
